File: app/src/main/java/tutorials/acviewmodel/home/ListViewModel.java
package tutorials.acviewmodel.home;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import java.util.List;

import javax.inject.Inject;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import tutorials.acviewmodel.model.Repo;
import tutorials.acviewmodel.networking.RepoApi;
import tutorials.acviewmodel.networking.RepoService;

public class ListViewModel extends ViewModel {

    private final MutableLiveData<List<Repo>> repos= new MutableLiveData<>();
    private final MutableLiveData<Boolean> repoLoadError= new MutableLiveData<>();
    private final MutableLiveData<Boolean> loading= new MutableLiveData<>();
    private final RepoService repoService;

    private Call<List<Repo>> repoCall;

    LiveData<List<Repo>> getRepos(){
        return  repos;
    }

    LiveData<Boolean> getError(){
        return repoLoadError;
    }

    LiveData<Boolean> getLoading(){
        return loading;
    }

    @Inject
    public ListViewModel(RepoService repoService){
        this.repoService = repoService;
        fetchRepos();
    }

    private void fetchRepos() {
        loading.setValue(true);
        repoCall = repoService.getRepositories();
        repoCall.enqueue(new Callback<List<Repo>>() {
            @Override
            public void onResponse(Call<List<Repo>> call, Response<List<Repo>> response) {
                repoLoadError.setValue(false);
                repos.setValue(response.body());
                loading.setValue(false);
                repoCall = null;
            }

            @Override
            public void onFailure(Call<List<Repo>> call, Throwable t) {
                repoLoadError.setValue(true);
                loading.setValue(false);
                repoCall = null;
            }
        });
    }

    @Override
    protected void onCleared() {
        if (repoCall != null) {
            repoCall.cancel();
            repoCall = null;
        }
    }
}

File: app/src/main/java/technologies/acviewmodel/screens/repos/viewmodel/ListViewModel.kt
package technologies.acviewmodel.screens.repos.viewmodel

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel

import javax.inject.Inject

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import technologies.acviewmodel.model.Repo
import technologies.acviewmodel.networking.RepoService

class ListViewModel @Inject
constructor(private val repoService: RepoService) : ViewModel() {

    private val repos = MutableLiveData<List<Repo>>()
    private val repoLoadError = MutableLiveData<Boolean>()
    private val loading = MutableLiveData<Boolean>()

    private var repoCall: Call<List<Repo>>? = null

    val error: LiveData<Boolean>
        get() = repoLoadError

    fun getRepos(): LiveData<List<Repo>> {
        return repos
    }

    fun getLoading(): LiveData<Boolean> {
        return loading
    }

    init {
        fetchRepos()
    }

    private fun fetchRepos() {
        loading.value = true
        repoCall = repoService.repositories
        repoCall!!.enqueue(object : Callback<List<Repo>> {
            override fun onResponse(call: Call<List<Repo>>, response: Response<List<Repo>>) {
                repoLoadError.value = false
                repos.value = response.body()
                loading.value = false
                repoCall = null
            }

            override fun onFailure(call: Call<List<Repo>>, t: Throwable) {
                repoLoadError.value = true
                loading.value = false
                repoCall = null
            }
        })
    }

    override fun onCleared() {
        if (repoCall != null) {
            repoCall!!.cancel()
            repoCall = null
        }
    }
}
